// ng-gorilla/esm2022/file-input/file-input-button.mjs
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class GrlFileInputButton {
    constructor() {
        this.grlFileInputAction = 'open';
        this._disabled = false;
    }
    get disabled() {
        return this._disabled || (this.grlFileInput && this.grlFileInput.disabled);
    }
    set disabled(value) {
        this._disabled = coerceBooleanProperty(value);
    }
    _click() {
        if (this.disabled) {
            return;
        }
        if (this.grlFileInputAction === 'open') {
            this.grlFileInput.open();
        }
        else if (this.grlFileInputAction === 'clear') {
            this.grlFileInput.clear();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.8", ngImport: i0, type: GrlFileInputButton, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.0.8", type: GrlFileInputButton, isStandalone: true, selector: "button[grlFileInputButtonFor]", inputs: { grlFileInput: ["grlFileInputButtonFor", "grlFileInput"], grlFileInputAction: "grlFileInputAction", disabled: "disabled" }, host: { listeners: { "click": "_click()" }, properties: { "attr.disabled": "disabled || null" } }, exportAs: ["grlFileInputButton", "grlFileInputButtonFor"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.8", ngImport: i0, type: GrlFileInputButton, decorators: [{
            type: Directive,
            args: [{
                    selector: 'button[grlFileInputButtonFor]',
                    exportAs: 'grlFileInputButton, grlFileInputButtonFor',
                    host: {
                        '(click)': '_click()',
                        '[attr.disabled]': 'disabled || null',
                    },
                    standalone: true,
                }]
        }], propDecorators: { grlFileInput: [{
                type: Input,
                args: ['grlFileInputButtonFor']
            }], grlFileInputAction: [{
                type: Input
            }], disabled: [{
                type: Input
            }] } });
